api/v1alpha1/customnbimage_types.go

```diff
@@ -24,69 +24,102 @@ import (
+
 const (
-	CNBiStrategyImageImport         = "import"
-	CNBiStrategyBuildUsingPython    = "build"
-	CNBiStrategyBuildUsingBaseImage = "baseImage"
+	// ImportImage will simply import the image from the given URL
+	ImportImage BuildType = "ImageImport"
+
+	// PackageList will build a custom image using a specific List of Python Packages
+	// if no RuntimeEnvironment is specified, a baseImage must be specified for the build
+	// TODO #70 this needs a validator to ensure that either a RuntimeEnvironment or a baseImage is specified
+	PackageList BuildType = "PackageList"
+
+	// BuildGitRepository will builds a custom image from a git repository
+	GitRepository BuildType = "GitRepository"
 )
 
-// CustomNBAnnotations is a list of annotations that are added to the custom notebook image
+// CNBi Annotations is a list of annotations that are added to the custom notebook image
 const (
-	CNBiAnnotationName        = "opendatahub.io/notebook-image-name"
-	CNBiAnnotationDescription = "opendatahub.io/notebook-image-desc"
-	CNBiAnnotationCreator     = "opendatahub.io/notebook-image-creator"
+	CNBiNameAnnotationKey        = "opendatahub.io/notebook-image-name"
+	CNBiDescriptionAnnotationKey = "opendatahub.io/notebook-image-desc"
+	CNBiCreatorAnnotationKey     = "opendatahub.io/notebook-image-creator"
 )
 
-// CustomNBImageStrategy is the strategy super-set of configurations for all strategies.
-type CustomNBImageStrategy struct {
-	// Type is the strategy
-	Type string `json:"type,omitempty"`
-	// From is the reference to the source image, used for import strategy
-	From string `json:"from,omitempty"`
+// BuildTypeSpec is the strategy super-set of configurations for all strategies.
+type BuildTypeSpec struct {
+	// BuildType is the strategy
+	// +required
+	// +kubebuilder:Required
+	BuildType BuildType `json:"buildType"`
+	// FromUrl is the reference to the source image, used for import strategy
+	// +optional
+	FromImage string `json:"fromImage,omitempty"`
+	// BuilderImage is the reference to the base image, used for building
+	// +optional
+	BuilderImage string `json:"builderImage,omitempty"`
+	// Repository is the URL of the git repository, used for building
+	// +optional
+	Repository string `json:"repository,omitempty"`
 }
 
 // CustomNBImageRuntimeSpec defines a Runtime Environment, aka 'the Python version used'
 type CustomNBImageRuntimeSpec struct {
 	// PythonVersion is the version of Python to use
+	// +optional
 	PythonVersion string `json:"pythonVersion,omitempty"`
 	// OSName is the Name of the Operating System to use
+	// +optional
 	OSName string `json:"osName,omitempty"`
 	// OSVersion is the Version of the Operating System to use
+	// +optional
 	OSVersion string `json:"osVersion,omitempty"`
-	// BaseImage is an alternative the the three above fields
-	BaseImage string `json:"baseImage,omitempty"`
+	// BuilderImage is an alternative the the three above fields
+	// +optional
+	BuilderImage string `json:"builderImage,omitempty"`
```

Comment:
Here we could (re)use the image reference field from the top level.

Regardless of having one or various, the semantics here are: either use a base image or a runtime spec - therefore, I believe the image reference should be at the same level as `runtimeEnvironment` rather than inside of it

Reply:
Should we just use `BaseImage` from the spec?